lib/vnet/dns/dns.go

```
@@ -0,0 +1,402 @@
+	ctx, cancel := context.WithTimeout(ctx, forwardRequestTimeout)
+	defer cancel()
+	deadline, _ := ctx.Deadline()
+
+	upstreamNameservers, err := s.upstreamNameserverSource.UpstreamNameservers(ctx)
+	if err != nil {
+		return trace.Wrap(err, "getting host default nameservers")
+	}
+	if len(upstreamNameservers) == 0 {
+		return trace.Errorf("no upstream nameservers")
+	}
+
+	// Forward the message to each upstream nameserver concurrently, the first to answer wins.
+	// Each goroutine will write a single error or a single response to the appropriate channel.
+	// Each goroutine should quickly exit after the context is canceled.
+	responses := make(chan []byte, len(upstreamNameservers))
+	errs := make(chan error, len(upstreamNameservers))
+	g, ctx := errgroup.WithContext(ctx)
+	ctx, cancel = context.WithCancel(ctx)
+	defer cancel()
+	for _, nameserver := range upstreamNameservers {
+		responseBuf, returnResponseBuf := s.getMessageBuffer()
+		defer returnResponseBuf()
+
+		nameserver := nameserver
+		g.Go(func() error {
+			slog := slog.With("nameserver", nameserver)
+			slog.DebugContext(ctx, "Forwarding request to upstream nameserver.")
+
+			upstreamConn, err := net.Dial("udp", nameserver)
+			if err != nil {
+				errs <- trace.Wrap(err, "dialing upstream nameserver")
+				return nil
+			}
+
+			// Immediately close the upstream conn after the context is canceled to unblock any i/o. This
+			// function will not return any answer until all errgroup goroutines have terminated.
+			go func() {
+				<-ctx.Done()
+				upstreamConn.Close()
+			}()
+
+			upstreamConn.SetDeadline(deadline)
+			_, err = upstreamConn.Write(buf)
+			if err != nil {
+				errs <- trace.Wrap(err, "writing message to upstream")
+				return nil
+			}
+			n, err := upstreamConn.Read(responseBuf)
+			if err != nil {
+				errs <- trace.Wrap(err, "reading forwarded DNS response")
+				return nil
+			}
```

Should either of these be a `err != nil && !utils.IsOKNetworkError(err)`?  If another server finishes, the context is cancelled and `upstreamConn.Close()` is called, thus I think these might error. (Not sure whether this piece has been manually tested or not at this point).

---

it's fine if these error though, if there's any success we return that and ignore errors from other upstreams